src/main/java/com/artipie/rpm/misc/StorageLocker.java

```diff
@@ -0,0 +1,139 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.rpm.misc;
+
+import com.artipie.asto.Key;
+import com.artipie.asto.Storage;
+import com.artipie.asto.blocking.BlockingStorage;
+import java.util.UUID;
+
+/**
+ * Locker for the storage.
+ * @since 0.9
+ * @todo #230:30min Use this class in RPM update to prohibit parallel update of the same repository:
+ *  add lock before starting the update and release it on terminate.
+ */
+public class StorageLocker {
+
+    /**
+     * Lock directory name.
+     */
+    private static final String FNAME = "lock-%s";
+
+    /**
+     * Lock directory name.
+     */
+    private static final String PTRN = "%s/lock-[a-z0-9-]{36}";
+
+    /**
+     * Storage.
+     */
+    private final Storage storage;
+
+    /**
+     * Repository key.
+     */
+    private final Key repo;
+
+    /**
+     * Lock file name.
+     */
+    private final String guid;
+
+    /**
+     * Ctor.
+     * @param storage Storage
+     * @param repo Repo key
+     */
+    public StorageLocker(final Storage storage, final Key repo) {
+        this.storage = storage;
+        this.repo = repo;
+        this.guid = UUID.randomUUID().toString();
+    }
+
+    /**
+     * Locks storage by adding random file to it.
+     * @throws IllegalStateException On error and if storage is already locked
+     */
+    void lock() {
+        final BlockingStorage bsto = new BlockingStorage(this.storage);
+        try {
+            if (this.noLocks(bsto)) {
+                synchronized (this.storage) {
+                    if (this.noLocks(bsto)) {
+                        bsto.save(
+                            new Key.From(this.repo, String.format(StorageLocker.FNAME, this.guid)),
+                            new byte[]{}
+                        );
+                    } else {
+                        throw this.error();
+                    }
+                }
+            } else {
+                throw this.error();
+            }
+        } catch (final InterruptedException ex) {
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    /**
+     * Releases lock by removing the lock file.
+     */
+    void release() {
```

Comment:
@olenagerasimova I guess it should be an async method, same as `lock`

Reply:
@olegmoz I do not think that lock should be obtained in async mode, that is the point of returning `ok` to the client and then saying that update is already running and failing?

Reply:
@olenagerasimova making application async is not about returning something to user, it's about not blocking threads for I/O operations. I thought that `Rpm` is designed in async way, so it requires async lock. If it is not the case, then making lock blocking is totally fine.